### ISSUE TITLE
chef-client crashes ruby with postgresql

### ISSUE THREAD

**Issue by Aketzu** [chef/chef#6485]
## Description

Running chef-client causes run to end up with ruby [BUG] Segmentation fault at 0x0000000000000000

## Chef Version

Crashes similarly on both chef-client 12.15.19 and 13.5.3

## Platform Version

Works OK in Debian 8.9, crashes after updating to Debian 9.2

PostgreSQL (and libpq) from PostgreSQL repo
`deb      "http://apt.postgresql.org/pub/repos/apt" stretch-pgdg main 9.6`

## Replication Case

Run `chef-client`. When trying minimal case it succeeds

## Client Output

Main points:
```
  * postgresql_database[artifactory] action createThe PGconn, PGresult, and PGError constants are deprecated, and will be
removed as of version 1.0.

You should use PG::Connection, PG::Result, and PG::Error instead, respectively.

Called from /var/chef/cache/cookbooks/database/libraries/provider_database_postgresql.rb:132:in `db'
/var/chef/cache/cookbooks/database/libraries/provider_database_postgresql.rb:132: [BUG] Segmentation fault at 0x0000000000000000
ruby 2.4.2p198 (2017-09-14 revision 59899) [x86_64-linux]

-- Control frame information -----------------------------------------------
c:0043 p:---- s:0216 e:000215 CFUNC  :initialize
c:0042 p:---- s:0213 e:000212 CFUNC  :new
c:0041 p:0203 s:0208 e:000207 METHOD /var/chef/cache/cookbooks/database/libraries/provider_database_postgresql.rb:132
c:0040 p:0044 s:0199 e:000198 METHOD /var/chef/cache/cookbooks/database/libraries/provider_database_postgresql.rb:97
c:0039 p:0008 s:0194 e:000193 LAMBDA /var/chef/cache/cookbooks/database/libraries/provider_database_postgresql.rb:46 [FINISH]

...

-- C level backtrace information -------------------------------------------
/opt/chef/embedded/lib/libruby.so.2.4(rb_vm_bugreport+0x559) [0x7f4067ed6e59] vm_dump.c:684
/opt/chef/embedded/lib/libruby.so.2.4(rb_bug_context+0xd0) [0x7f4067d6e1d0] error.c:506
/opt/chef/embedded/lib/libruby.so.2.4(sigsegv+0x3e) [0x7f4067e4f93e] signal.c:907
/lib/x86_64-linux-gnu/libpthread.so.0 [0x7f4067ac50c0]
/lib/x86_64-linux-gnu/libc.so.6 [0x7f4066ef6386]
/opt/chef/embedded/lib/libcrypto.so.1.0.0(getrn+0x89) [0x7f40652c7339]
/opt/chef/embedded/lib/libcrypto.so.1.0.0(lh_insert+0x42) [0x7f40652c7602]
/opt/chef/embedded/lib/libcrypto.so.1.0.0(OBJ_NAME_add+0x85) [0x7f4065223625]
/usr/lib/x86_64-linux-gnu/libcrypto.so.1.1 [0x7f405b3dca3c]
/usr/lib/x86_64-linux-gnu/libcrypto.so.1.1 [0x7f405b3f40d9]
/lib/x86_64-linux-gnu/libpthread.so.0 [0x7f4067ac2739]
/usr/lib/x86_64-linux-gnu/libcrypto.so.1.1(CRYPTO_THREAD_run_once+0x9) [0x7f405b449f59]
/usr/lib/x86_64-linux-gnu/libcrypto.so.1.1(OPENSSL_init_crypto+0x1b3) [0x7f405b3f4533]
/usr/lib/x86_64-linux-gnu/libssl.so.1.1(OPENSSL_init_ssl+0x74) [0x7f405b74bbd4]
/usr/lib/x86_64-linux-gnu/libpq.so.5 [0x7f405b9b1a8c]
/usr/lib/x86_64-linux-gnu/libpq.so.5(PQconnectPoll+0xbbe) [0x7f405b999abe]
/usr/lib/x86_64-linux-gnu/libpq.so.5 [0x7f405b99a57a]
/usr/lib/x86_64-linux-gnu/libpq.so.5(PQconnectdb+0x27) [0x7f405b99d867]
/opt/chef/embedded/lib/ruby/gems/2.4.0/gems/pg-0.21.0/lib/pg_ext.so(gvl_PQconnectdb_skeleton+0xc) [0x7f405bbdc70c] gvl_wrappers.c:9
/opt/chef/embedded/lib/libruby.so.2.4(rb_thread_call_without_gvl+0x3b) [0x7f4067e8f5db] thread.c:1320
/opt/chef/embedded/lib/ruby/gems/2.4.0/gems/pg-0.21.0/lib/pg_ext.so(gvl_PQconnectdb+0x29) [0x7f405bbdcb19] gvl_wrappers.c:10
/opt/chef/embedded/lib/ruby/gems/2.4.0/gems/pg-0.21.0/lib/pg_ext.so(pgconn_init+0x5a) [0x7f405bbe513a] pg_connection.c:283
```

Full normal output
https://gist.github.com/Aketzu/dfc4d672245bf41988080344395db82e

## Cookbook

Around the crash our vilant-artifactory cookbook is executing the following
```
postgresql_connection_info = { host:  'localhost',
                               port:  node['postgresql']['config']['port'],
                               username: 'postgres',
                               password: node['postgresql']['password']['postgres'] }

node['vilant']['databases'].each do |_, settings|
  postgresql_database settings['database_database'] do
    connection postgresql_connection_info
    action :create
  end
...
```

provider_database_postgresql.rb:132 contains ```::PGconn.new(...)```
https://github.com/chef-boneyard/database/blob/master/libraries/provider_database_postgresql.rb#L132

If I try to run same ```::PGconn.new``` separately with /opt/chef/embedded/ruby everything works OK and it doesn't crash.

## Toughts

Everything worked fine before I did Debian 9 update. In C backtrace above there is Debian libcrypto.so.1.1 and Chef libcrypto.so.1.0.0 mixed so that might be causing the issue.

libpq.so.5 is provided by postgresql.org public apt repo. It is linked against (system) libssl.so.1.1 and libcrypto.so.1.1

/opt/chef/embedded/lib/ruby/gems/2.4.0/gems/pg-0.21.0/lib/pg_ext.so is also linked against libssl.so.1.1 and libcrypto.so.1.1

libcrypto.so.1.1 naturally doesn't link against chef libcrypto.1.0.0 but somehow ends up calling older chef library instead of itself. Almost smells like monkeypatching...

**Comment by Aketzu** [comment]
Here's a minimal testcase that crashes
```ruby
#!/opt/chef/embedded/bin/ruby
require 'openssl'
require 'pg'
::PGconn.new( host: 'localhost', port: 22 )
```
(Any open TCP port works, no need to have postgresql-server installed)

**Comment by lamont-granquist** [comment]
> there is Debian libcrypto.so.1.1 and Chef libcrypto.so.1.0.0 mixed so that might be causing the issue.

yeah, your postgres c-libraries need to be linked against the libcrypto that /opt/chef/embedded/bin/ruby is linked against and that we ship under /opt/chef/embedded/lib.

this is not a bug in this repo.

you need to be using the postgresql::ruby code to install the pg libs into the omnibus and linked against the omnibus binaries:

https://github.com/sous-chefs/postgresql/blob/master/recipes/ruby.rb#L64-L124

it looks like that only gets exercised if the attempt to chef_gem install fails, which smells wrong to me because if you have pg libs installed in the system then it'll link the pg gem against those, which will then transitively pick up the openssl libraries, and that or something like that might be happening.

there's an open RFC that needs to get picked up again to include the pg libraries and the pg gem in the omnibus binaries because this is such a pain:

https://github.com/chef/chef-rfc/blob/master/rfc063-omnibus-chef-native-gems.md

that got derailed for a long time because we had some horrible build issues, then i got derailed by other projects, so that has fallen off my radar entirely (although nokogiri did get added to the build).

**Comment by Aketzu** [comment]
Made two different ways to fix this.

### Build chef-client with OpenSSL 1.1
Requires chef/omnibus-software#889 and changing openssl to "1.1.0f" in chef/omnibus_overrides.rb.
Works in my Debian 9 but then probably breaks on older platforms. If you are ever going to make Debian 9 build then this is probably a good idea anyways

### Package libpq into chef
Implements part of RFC063 as you mentioned. Work done in chef/omnibus-software#890
Then just needs `dependency "pg-gem"` in chef/omnibus/config/projects/chef.rb

**Comment by lamont-granquist** [comment]
Yes, the solution of "build-client with OpenSSL 1.1" just pushes the problem around.  That will make it work on distros that package 1.1, but it will now fail on distros with 1.0.1/1.02 or (0.9.6 or whatever shipped on RHEL5).  This is not something that I'd advise any client to do since it involves attempting to build omnibus itself.  If you can do that, it's sort of cool that you've figured it out, but you're very much off the reservation at that point (our helpdesk support will most likely be pretty useless to you at that point).

Since RFC063 has gotten unblocked I need to get around back onto that, but other emergencies are in the way.

The postgresql cookbook is a better approach, and if I'm right that it looks like there's a bug in there, then that bug should really get reported and fixed (although it's odd to me that it hasn't ever been reported before and that is like a 4-5 year old bug, but that happens sometimes...).  It can be validated that there's a bug there just by forcing the exception handler that does all the compilation to always run and avoiding the install-chef-gem-and-rescue codepath entirely.

**Comment by lamont-granquist** [comment]
closing this in favor of sous-chefs/postgresql#462
